### PR TITLE
[MLv2] [FE] Filter modal segments

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/34112

Changes:
- Adds segments to the filter modal. They are available in the first group (tab).
- Make sure they can be filtered by using the top-level search in the modal. 
- Converts multi-value pickers to `MultiSelect` from `mantine`. It's not styled yet, this is TBD separately. Also handling field values TBD in a separate PR.

How to verify:
- You can add a segment as a filter and remove it from the query. Query should be modified and work correctly.

Screenshot 2023-12-12 at 16 23 00

### CODE REVIEW

File: frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx [hunk 1]
```diff
@@ -85,23 +84,26 @@ export function CoordinateFilterEditor({
...
   onChange,
 }: NumberValueInputProps) {
   if (hasMultipleValues) {
     return (
-      <FilterValuePicker value={values} column={column} onChange={onChange} />
+      <FilterValuePicker
+        values={values.map(value => String(value))}
+        placeholder={t`Enter a number`}
+        getCreateLabel={query => (isFinite(Number(query)) ? query : null)}
+        onChange={values => onChange(values.map(value => Number(value)))}
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx:105]: `Number("") === 0`, it's possible to enter whitespace and it will be converted to 0.
Previously spaces were not converted to zeros.

I think it'd be better to use `parseInt` and filter out `NaN`s (in `onChange`).

File: frontend/src/metabase/ui/components/inputs/index.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ export * from "./DateInput";
 export * from "./DatePicker";
 export * from "./FileInput";
 export * from "./Input";
+export * from "./MultiSelect";
```
Comment (kamilmielnik) [inline frontend/src/metabase/ui/components/inputs/index.ts:7]: Using the same value twice results in the following React warning:

```
Warning: Encountered two children with the same key, `9`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
```

image

Previously it was not possible to enter the same value twice.

File: frontend/src/metabase/querying/components/FilterModal/utils/filters.ts [hunk 3]
```diff
@@ -63,6 +78,37 @@ export function removeFilters(query: Lib.Query) {
...
+  return segmentItems.reduce((query, { segment, stageIndex }) => {
+    return Lib.filter(query, stageIndex, segment);
+  }, query);
+}
+
+export function removeSegmentFilters(
+  query: Lib.Query,
+  segmentItems: SegmentItem[],
+) {
+  const filterGroups = segmentItems.flatMap(
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterModal/utils/filters.ts:94]:
```suggestion
  const filterGroups = segmentItems.map(
```

File: frontend/src/metabase/querying/components/FilterModal/utils/search.ts [hunk 4]
```diff
@@ -11,19 +11,29 @@ export function searchGroupItems(
   searchText: string,
 ): GroupItem[] {
   const searchValue = searchText.toLowerCase();
+  const searchValueHasSegments = t`segments`.includes(searchValue);
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterModal/utils/search.ts:14]: It took me a while to figure out why the value isn't ```searchValue.includes(t`segments`)```, but it was the variable name that was confusing.

```suggestion
  const isSearchingForSegments = t`segments`.includes(searchValue);
```

File: frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx [hunk 5]
```diff
@@ -85,23 +84,26 @@ export function CoordinateFilterEditor({
...
-  column,
   values,
   valueCount,
   hasMultipleValues,
   onChange,
 }: NumberValueInputProps) {
   if (hasMultipleValues) {
     return (
-      <FilterValuePicker value={values} column={column} onChange={onChange} />
+      <FilterValuePicker
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx:101]: Coordinates used to be formatted like this:

image

Now they're not.
Is this going to be handled in a separate PR?

File: frontend/src/metabase/querying/components/FilterModal/SegmentFilterEditor/SegmentFilterEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import { useMemo } from "react";
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterModal/SegmentFilterEditor/SegmentFilterEditor.tsx:1]: I removed a segment and went back to the query builder - the removed segment was still available in the filter modal (until I refreshed a page).

In `master` this does not happen.

File: frontend/src/metabase/ui/components/inputs/index.ts (hunk 2 above)
Comment (ranquild) [reply frontend/src/metabase/ui/components/inputs/index.ts:7]: https://github.com/metabase/metabase/issues/36748

File: frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx (hunk 5 above)
Comment (ranquild) [reply frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx:101]: https://github.com/metabase/metabase/issues/36749

File: frontend/src/metabase/querying/components/FilterModal/utils/filters.ts (hunk 3 above)
Comment (ranquild) [reply frontend/src/metabase/querying/components/FilterModal/utils/filters.ts]: Fixed

File: frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx (hunk 1 above)
Comment (ranquild) [reply frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx]: Fixed

**Comment by ranquild** [comment]
@kamilmielnik I've addressed simple things and create separate issues for bigger issues. Please take a look

File: frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx [hunk 7]
```diff
@@ -85,23 +84,26 @@ export function CoordinateFilterEditor({
...
+        values={values.map(value => String(value))}
+        placeholder={t`Enter a number`}
+        getCreateLabel={query => (isFinite(parseFloat(query)) ? query : null)}
+        onChange={values => onChange(values.map(value => Number(value)))}
```
Comment (kamilmielnik) [inline frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx:105]:
```suggestion
        onChange={values => onChange(values.map(value => parseFloat(value)).filter(value => isFinite(value)))}
```

Same comment for 3 other places

**Review by kamilmielnik — APPROVED** [review]

File: frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx (hunk 7 above)
Comment (ranquild) [reply frontend/src/metabase/querying/components/FilterModal/CoordinateFilterEditor/CoordinateFilterEditor.tsx:105]: I'll address that in the next PR where the component is changed
